include/triton/Dialect/Gluon/Transforms/InferLayoutUtils.h

```diff
@@ -0,0 +1,32 @@
+#ifndef TRITON_DIALECT_GLUON_TRANSFORMS_INFERLAYOUTUTILS_H_
+#define TRITON_DIALECT_GLUON_TRANSFORMS_INFERLAYOUTUTILS_H_
+
+#include "triton/Dialect/Gluon/IR/Dialect.h"
+#include "triton/Dialect/Gluon/Transforms/Passes.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/PriorityWorklist.h"
+
+namespace mlir::triton::gluon {
+
+struct LayoutInfo {
+  Attribute encoding;
+  // Some operations can infer one of many encodings,
+  // we model this by setting the mayVary flag on encodings
+  // derived from these ops.
+  // If "may vary" is set then we allow conflicts, and when
+  // resolving conflicts we prefer encodings that are not allowed to vary.
+  bool mayVary = false;
+
+  operator bool() { return bool(encoding); }
+};
```

Comment:
This doesn't need to be public any more, you can move into `InferLayoutUtils.cpp`